### PR TITLE
Checkpoint closing files

### PR DESCRIPTION
**Pull request by michaelcahill**
These are fixes for SERVER-17587.  The basic idea is to rely on meta tracking to take care of flushing the metadata during any operation that needs to close a dirty tree.

Note that this change relies on `__wt_meta_track_off` making the metadata stable.  That function checkpoints the metadata if logging is not enabled.  It is sufficient to fix the included test case, but deserves careful review to determine whether it does enough to make the metadata stable on disk before old checkpoints are freed in general.  (I suspect it doesn't, today, but keep in mind that we're closing a single tree for some exclusive operation such as verify -- there is no concurrent eviction).

### CODE REVIEW

File: src/include/btree.i [hunk 1]
```diff
@@ -970,7 +970,7 @@ __wt_page_can_evict(WT_SESSION_IMPL *session, WT_PAGE *page, int check_splits)
 	 * a transaction value, once that's globally visible, we know we can
 	 * evict the created page.
 	 */
-	if (WT_PAGE_IS_INTERNAL(page) &&
+	if (check_splits && WT_PAGE_IS_INTERNAL(page) &&
```
Comment (michaelcahill) [inline src/include/btree.i]: This part of the change deserves some explanation and thought.  I don't know for sure that this is necessary, but I'm now asserting that all pages in a clean tree can be safely discarded, and this is the one case I'm not sure is guaranteed anywhere.

We could include `mod->mod_split_txn` in the calculation of `btree->rec_max_txn`, maybe?

File: test/suite/test_sweep01.py [hunk 2]
```diff
@@ -42,10 +42,7 @@ class test_sweep01(wttest.WiredTigerTestCase, suite_subprocess):
     uri = 'table:' + tablebase
     numfiles = 50
     numkv = 1000
-    ckpt_list = [
-        ('off', dict(ckpt=0)),
-        ('on', dict(ckpt=10)),
-    ]
+    ckpt=10
```
Comment (michaelcahill) [inline test/suite/test_sweep01.py]: This test will no longer pass without checkpoints, because dirty files will not be swept.

**Comment by agorrod** [comment]
@michaelcahill I've gone through this change and it makes sense to me. @keithbostic should review also before merging.

File: src/btree/bt_sync.c [hunk 3]
```diff
@@ -153,6 +157,9 @@ __sync_file(WT_SESSION_IMPL *session, int syncop)
 				++leaf_pages;
 			}
 			WT_ERR(__wt_reconcile(session, walk, NULL, 0));
+
+			if (TXNID_LT(btree->rec_max_txn, mod->rec_max_txn))
+				btree->rec_max_txn = mod->rec_max_txn;
```
Comment (keithbostic) [inline src/btree/bt_sync.c]: I'd put the common update code in one place:
```
118a119,122
>                       if (mod != NULL &&
>                           TXNID_LT(btree->rec_max_txn, mod->rec_max_txn))
>                               btree->rec_max_txn = mod->rec_max_txn;
>
120,123c124
<                       if (!__wt_page_is_modified(page)) {
<                               if (mod != NULL && TXNID_LT(
<                                   btree->rec_max_txn, mod->rec_max_txn))
<                                       btree->rec_max_txn = mod->rec_max_txn;
---
>                       if (!__wt_page_is_modified(page))
125d125
<                       }
160,162d159
<
<                       if (TXNID_LT(btree->rec_max_txn, mod->rec_max_txn))
<                               btree->rec_max_txn = mod->rec_max_txn;
```
and let the compiler do the work.

It also would add a natural place to add a comment explaining what's happening here, I think that's worth doing regardless.

File: src/include/btree.i (hunk 1 above)
Comment (keithbostic) [inline src/include/btree.i]: `mod_split_txn` is only set when we deepen the tree, and the child pages we create when we deepen the tree dirty the tree, so I don't think this is necessary.

**Comment by keithbostic** [comment]
@michaelcahill, lgtm, let's merge.

File: src/btree/bt_sync.c (hunk 3 above)
Comment (michaelcahill) [inline src/btree/bt_sync.c]: That change won't deal with dirty pages that are written (and become clean) during the checkpoint.  I'll add comments regardless.

**Pull request merged**